invoke-ai/InvokeAI
# feat(ui): add support for custom field types

## What type of PR is this? (check all applicable)



## Have you discussed this change with the InvokeAI team?

## Description

Node authors may now create their own arbitrary/custom field types. Any pydantic model is supported.

Two notes:
1. Your field type's class name must be unique.

Suggest prefixing fields with something related to the node pack as a kind of namespace.

2. Custom field types function as connection-only fields.

For example, if your custom field has string attributes, you will not get a text input for that attribute when you give a node a field with your custom type.

This is the same behaviour as other complex fields that don't have custom UIs in the workflow editor - like, say, a string collection.

## QA Instructions, Screenshots, Recordings

Clone this repo into your `invokeai/nodes/` dir: https://github.com/psychedelicious/test-arbitrary-fields

It contains two nodes:
- `Custom Field Test 1`
- `Custom Field Test 2`

These two nodes use a custom field called `CustomField`. 

Try adding those nodes and connecting them. Should work like any node.
